### PR TITLE
Make Plutarch sound

### PR DESCRIPTION
**Pull request by L-as**
All new way of deriving `PlutusType` and all other `P`-classes.
`pcon` is always strict (`plazy` hasn't been added yet).
Lots of general clean-ups, removing unsound code, etc.
`PForall`, `PSome`, `PFix`, instead of extra type parameter on `PInner`.
Fixed upcasting and downcasting by adding magical `P{Co{,ntra}v,V}ariant` type families.
Added a lot more warnings by default.
Removed AllowAmbiguousTypes in a lot of places.
Made FFI a lot simpler.
Removed a lot of bad uses of overlapping instances.

This is the final pressing stuff for 1.2 AFAICT. The tests don't work yet, and plutarch-extra also needs to be fixed.

### CODE REVIEW

File: plutarch-test/plutarch-base/Plutarch/PlutusTypeSpec.hs [hunk 1]
```diff
@@ -289,23 +288,9 @@ describe "sanity checks" $ do
...
-  internally.
--}
-instance PlutusType AB where
-  type PInner AB _ = PInteger
-
-  pcon' A = 0
-  pcon' B = 1
-
-  pmatch' x f =
-    pif (x #== 0) (f A) (f B)
```
Comment (MangoIV) [inline plutarch-test/plutarch-base/Plutarch/PlutusTypeSpec.hs:308]: Why did you remove this?

**Review by MangoIV — CHANGES_REQUESTED** [review]
I will have to test some stuff I didn't understand from reading. This really needs documentation if it's going to be used as a library. This should be priority

File: Plutarch/Bool.hs [hunk 2]
```diff
@@ -84,7 +80,11 @@ infix 4 #==
 
 class PEq t => POrd t where
   (#<=) :: Term s t -> Term s t -> Term s PBool
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: That would destroy the point?

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: or do I misunderstand how it works? Can you make it err if you either implement all or none?

File: Plutarch/Builtin.hs [hunk 3]
```diff
@@ -186,19 +206,30 @@ instance PListLike PBuiltinList where
...
-  xs #== ys = pdata xs #== pdata ys
+instance (PLift a, PEq a) => Fc 'False a where
+  fc _ xs ys = plistEquals # xs # ys
 
-instance {-# OVERLAPPING #-} PEq (PBuiltinList PData) where
-  xs #== ys = pdata xs #== pdata ys
+instance PIsData (PBuiltinList a) => Fc 'True a where
+  fc _ xs ys = pdata xs #== pdata ys
+
+instance Fc (F a) a => PEq (PBuiltinList a) where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Builtin.hs:223]: Thanks

File: Plutarch/Builtin.hs [hunk 4]
```diff
@@ -264,16 +297,16 @@ pforgetData = pupcast
...
-pforgetData' :: forall a (p :: PType -> PType) s. Proxy p -> Term s (p (PAsData a)) -> Term s (p PData)
-pforgetData' _ = punsafeCoerce
+pforgetData' :: forall a (p :: PType -> PType) s. PCovariant p => Proxy p -> Term s (p (PAsData a)) -> Term s (p PData)
+pforgetData' = pupcastF
 
 -- | Inverse of 'pforgetData''.
-prememberData :: forall (p :: PType -> PType) s. Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
-prememberData Proxy = punsafeCoerce
+prememberData :: forall (p :: PType -> PType) s. PVariant p => Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
+prememberData Proxy = let _ = witness (Proxy @(PVariant p)) in punsafeCoerce
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Builtin.hs:305]: Yep

File: Plutarch/Builtin.hs [hunk 5]
```diff
@@ -282,7 +315,12 @@ prememberData' Proxy = punsafeCoerce
 -}
 class PIsData a where
   pfromDataImpl :: Term s (PAsData a) -> Term s a
+  default pfromDataImpl :: PIsData (PInner a) => Term s (PAsData a) -> Term s a
+  pfromDataImpl x = punsafeDowncast $ pfromDataImpl $ (punsafeCoerce x :: Term _ (PAsData (PInner a)))
+
   pdataImpl :: Term s a -> Term s PData
+  default pdataImpl :: PIsData (PInner a) => Term s a -> Term s PData
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Builtin.hs:322]: You can't implement that without `punsafeCoerce`.

File: Plutarch/Builtin.hs [hunk 6]
```diff
@@ -420,12 +451,9 @@ Example:
 > pconstantData @PInteger 42
 -}
 pconstantData :: forall p h s. (ToData h, PLifted p ~ h, PConstanted h ~ p) => h -> Term s (PAsData p)
-pconstantData x = punsafeCoerce $ pconstant $ PlutusTx.toData x
-
-newtype Flip f a b = Flip (f b a)
+pconstantData x = let _ = witness (Proxy @(PLifted p ~ h, PConstanted h ~ p)) in punsafeCoerce $ pconstant $ PlutusTx.toData x
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Builtin.hs:454]: I don't think so.

File: Plutarch/DataRepr/Internal.hs [hunk 7]
```diff
@@ -183,13 +178,14 @@ instance (POrd x, PIsData x) => POrd (PDataRecord '[label ':= x]) where
...
     pure $ pfromData x #< pfromData y
+
   l1 #<= l2 = unTermCont $ do
     PDCons x _ <- tcont $ pmatch l1
     PDCons y _ <- tcont $ pmatch l2
 
     pure $ pfromData x #<= pfromData y
 
-instance (POrd x, PIsData x, POrd (PDataRecord (x' ': xs))) => POrd (PDataRecord ((label ':= x) ': x' ': xs)) where
+instance (SListI xs, POrd x, PIsData x, POrd (PDataRecord (x' ': xs))) => POrd (PDataRecord ((label ':= x) ': x' ': xs)) where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/DataRepr/Internal.hs:188]: Basically yeah, unfortunately. You could have a more relaxed constraint that only needs to know the head I suppose, but I didn't see the point.

File: Plutarch/Internal/Newtype.hs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  PlutusTypeStrat,
+  PlutusTypeStratConstraint,
+  derivedPCon,
+  derivedPMatch,
+ )
+
+data PlutusTypeNewtype
+
+class (PGeneric a, PCode a ~ '[ '[GetPNewtype a]]) => Helper (a :: PType)
+instance (PGeneric a, PCode a ~ '[ '[GetPNewtype a]]) => Helper (a :: PType)
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Internal/Newtype.hs:20]: IIRC because the type family arity is different

File: Plutarch/Internal/Quantification.hs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+module Plutarch.Internal.Quantification (PForall (PForall), PSome (PSome), PFix (PFix)) where
+
+import Plutarch.Internal (PType, Term)
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Internal/Quantification.hs:3]: I don't think so. It's obvious from the definitions.

File: Plutarch/Lift.hs [hunk 10]
```diff
@@ -208,6 +207,7 @@ class ToBuiltin' a arep | a -> arep where
 class FromBuiltin' arep a | arep -> a where
   fromBuiltin' :: arep -> a
 
+-- FIXME this overlappable instance is nonsense and disregards the fundep
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Lift.hs:210]: It's hard. They are used.

File: Plutarch/Reducible.hs [hunk 11]
```diff
@@ -1,17 +1,37 @@
...
+ Given a type family @F : A -> Type@, you can make the following
+ @
+   type F' :: A -> Type
+   newtype F' (a :: A) = F' (NoReduce (F a)) deriving stock Generic
+ @
+ It is then true that @forall a. Reduce (F' a) ~ F a@.
+-}
+type family Reduce (x :: Type) :: Type where
+  Reduce (NoReduce a) = a
+  Reduce (Term s a) = Term s a -- FIXME remove
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Reducible.hs:32]: It's a work-around because `Term` doesn't implement `Generic`.

File: plutarch-test/plutarch-base/Plutarch/PlutusTypeSpec.hs (hunk 1 above)
Comment (L-as) [reply plutarch-test/plutarch-base/Plutarch/PlutusTypeSpec.hs:308]: Because it's not a good example anymore! You're never supposed to implement `PlutusType` yourself. In fact, `pcon'` and `pmatch'` aren't even exported anymore I think!

File: Plutarch/Reducible.hs [hunk 12]
```diff
@@ -1,17 +1,37 @@
...
-instance Reducible a => Reducible (Const a b) where
-  type Reduce (Const a b) = Reduce a
+ Given a type family @F : A -> Type@, you can make the following
+ @
+   type F' :: A -> Type
+   newtype F' (a :: A) = F' (NoReduce (F a)) deriving stock Generic
+ @
+ It is then true that @forall a. Reduce (F' a) ~ F a@.
+-}
+type family Reduce (x :: Type) :: Type where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Reducible.hs:30]: No, this is very different

**Comment by L-as** [comment]
@MangoIV why not just do it in a separate PR

**Comment by MangoIV** [comment]
I guess that's fair, then we open an issue and add it to plutarch v1.2 todo list maybe?

File: Plutarch/DataRepr/Internal/Field.hs [hunk 13]
```diff
@@ -220,45 +212,45 @@ class BindFields (ps :: [PLabeledType]) (bs :: [ToBind]) where
   --
   --    A continuation is returned to enable sharing of
   --    the generated bound-variables.
-  bindFields :: Term s (PDataRecord ps) -> TermCont s (HRec (BoundTerms ps bs s))
+  bindFields :: Proxy bs -> Term s (PDataRecord ps) -> TermCont s (HRec (BoundTerms ps bs s))
 
-instance {-# OVERLAPPING #-} BindFields ((l ':= p) ': ps) ( 'Bind ': '[]) where
-  bindFields t =
+instance {-# OVERLAPPABLE #-} BindFields ((l ':= p) ': ps) ( 'Bind ': '[]) where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/DataRepr/Internal/Field.hs:217]: In this case you can just remove the `OVERLAPPABLE`, but `OVERLAPPING` and `OVERLAPS` aren't safe pragmas.

File: Plutarch/Bool.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: The only fix I can think of besides fixing `MINIMAL` is avoiding default implementations and doing this:
```haskell
class DeriveOrd a
instance {-# OVERLAPPABLE #-} DeriveOrd a => Ord a where
  ...
```
Then you could do `deriving anyclass DeriveOrd`.
This however doesn't satisfy me because:
- Haskell doesn't warn on incoherent usage of this overlappable instance.
- Even if it did, you can still implement both `DeriveOrd` *and* `Ord` manually!

What we really want is
```haskell
instance DeriveOrd a => Ord a where
  ...
```
which should disallow manually deriving `Ord A` if `DeriveOrd A`.
You could implement this by taking constraints into account when resolving instances possibly.
If you have `[W] Ord A` and you have both the instances `DeriveOrd A => Ord A` and `Ord A`, GHC could discard the first one by noting that the constraint isn't fulfilled. This calls for another GHC proposal.

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: Got a great answer by tomjaguarpaw, but it means we have to have only one method per class... huge PITA

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: and also breaking change

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: Perhaps the OVERLAPPABLE approach is better?

[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Bool.hs:82]: I honestly think it's fine - no need to make it more complicated.

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: The issue is this can lead to subtle security bugs. Implementing just `#<=` while `#<` gives a completely incompatible resultl.

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: hmm

[the remark this answers is not in the document]
Comment (TotallyNotChase) [reply Plutarch/Bool.hs:82]: this is why we have typeclass laws, yes?

[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Bool.hs:82]: Typeclass laws don't warn you about anything...

File: Plutarch/Num.hs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  pthrow,
+  punsafeCoerce,
+  (:-->),
+ )
+import Plutarch.Internal.Other (pto)
+import Plutarch.Internal.PLam ((#))
+import Plutarch.Internal.PlutusType (PInner)
+import Plutarch.Unsafe (punsafeDowncast)
+
+class PNum (a :: PType) where
```
Comment (TotallyNotChase) [inline Plutarch/Num.hs:17]: you know, I'm still unsure if adding a base-like `Num` is a good idea. It'd be nice to just have `Num` instances as we do already and focus on a nice, convenient numerical hierarchy instead.

this class might add backwards compat burden later on

File: Plutarch/Builtin.hs [hunk 15]
```diff
@@ -264,16 +297,16 @@ pforgetData = pupcast
...
-prememberData :: forall (p :: PType -> PType) s. Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
-prememberData Proxy = punsafeCoerce
+prememberData :: forall (p :: PType -> PType) s. PVariant p => Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
+prememberData Proxy = let _ = witness (Proxy @(PVariant p)) in punsafeCoerce
 
 -- | Like 'prememberData' but generalised.
-prememberData' :: forall a (p :: PType -> PType) s. PSubtype PData a => Proxy p -> Term s (p a) -> Term s (p (PAsData a))
-prememberData' Proxy = punsafeCoerce
+prememberData' :: forall a (p :: PType -> PType) s. (PSubtype PData a, PVariant p) => Proxy p -> Term s (p a) -> Term s (p (PAsData a))
+prememberData' Proxy = let _ = witness (Proxy @(PSubtype PData a, PVariant p)) in punsafeCoerce
```
Comment (TotallyNotChase) [inline Plutarch/Builtin.hs:309]: do `PIsDataRepr` instances have a `PSubtype` relationship with `PData`?

File: Plutarch/Internal/PlutusType.hs [hunk 16]
```diff
@@ -1,317 +1,125 @@
...
-  , ToPType xs ~ IndexList n (PCode s a)
-  , GPMatch a (n + 1) xss c s
-  , PLamL (ToPType xs) c s
-  , MkNS n (PCode s a) (NP (Term s))
-  , All Top (ToPType xs)
-  , All Top xs
+import Plutarch.Internal.Quantification (PFix (PFix), PForall (PForall), PSome (PSome))
+import Plutarch.Internal.Witness (witness)
+
+class PlutusTypeStrat (strategy :: Type) where
```
Comment (TotallyNotChase) [inline Plutarch/Internal/PlutusType.hs:39]: what was the reason behind choosing making the `strategy` open kinded? is there something potentially bad with using data kinds or are you just leaving room for expansion by users?

File: Plutarch/DataRepr/Internal/Field.hs [hunk 17]
```diff
@@ -68,14 +68,16 @@ import Plutarch.DataRepr.Internal.HList (
...
  )
-import Plutarch.Internal (punsafeCoerce)
+import Plutarch.Internal.Witness (witness)
 import Plutarch.TermCont (TermCont (TermCont), runTermCont)
 
 --------------------------------------------------------------------------------
 ---------- PDataField class & deriving utils
 
+type family Helper (x :: PType) :: [PLabeledType] where
+  Helper (PDataSum '[y]) = y
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/DataRepr/Internal/Field.hs:79]: This isn't a singleton. I opted not to rename it.

File: Plutarch/Internal/PLam.hs [hunk 18]
```diff
@@ -52,10 +52,10 @@ infixr 0 #$
 class PLamN (a :: Type) (b :: PType) (s :: S) | a -> b, s b -> a where
   plam :: forall c. (Term s c -> a) -> Term s (c :--> b)
 
-instance (a' ~ Term s a) => PLamN a' a s where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Internal/PLam.hs:55]: OVERLAPPING is incoherent

File: Plutarch/Internal/PlutusType.hs [hunk 19]
```diff
@@ -1,317 +1,125 @@
...
-{- |
-  Like `ScottFn`, but without the PDelayed case.
--}
-type ScottFn' :: [PType] -> PType -> PType
-type family ScottFn' xs b where
-  ScottFn' '[] b = b
-  ScottFn' (x ': xs) b = x :--> ScottFn' xs b
+  type DPTStrat a :: Type
+
+class PlutusType (a :: PType) where
```
[the remark this answers is not in the document]
Comment (L-as) [reply Plutarch/Internal/PlutusType.hs:55]: Separate PR!

File: Plutarch/Internal/PlutusType.hs (hunk 16 above)
Comment (L-as) [reply Plutarch/Internal/PlutusType.hs:39]: There's no reason to make it closed, and it would make it more complex.

File: Plutarch/Builtin.hs (hunk 15 above)
Comment (L-as) [reply Plutarch/Builtin.hs:309]: That doesn't exist anymore.

File: Plutarch/Num.hs (hunk 14 above)
Comment (L-as) [reply Plutarch/Num.hs:17]: The reason to have this isn't to have a numerical hierarchy. It's simply to allow deriving `Num`.

File: Plutarch/Builtin.hs (hunk 15 above)
Comment (TotallyNotChase) [reply Plutarch/Builtin.hs:309]: ah, I see. Well, is there such a subtype relationship between `PlutusType` derived via `PlutusTypeData` then?

Comment (L-as) [reply Plutarch/Builtin.hs:309]: Yep

File: Plutarch/Num.hs (hunk 14 above)
Comment (TotallyNotChase) [reply Plutarch/Num.hs:17]: should `Semigroup`/`Monoid` get the same treatment?

Comment (L-as) [reply Plutarch/Num.hs:17]: Unfortunately yeah, I'm not sure there's a way to do this for all such typeclasses. In general, `C (Term s A)` is an orphan instance, so it's not great style in the first place. For all such `C`, it seems like we want `PC a => C (Term s a)`
